### PR TITLE
Skeleton for testing http query and form parameters

### PR DESCRIPTION
**Pull request by maramihali**
This is the simple skeleton of my http request tester. I wanted to submit it in an early stage to get some feedback on the structure. I added some comments to the code whose behaviour is a bit confusing to me.

### CODE REVIEW

File: http_server/request_testing/request_listener.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	connChannel chan net.Conn
+	serverEndpoint net.Conn
+}
+
+// For each request, we create an HTTP listener that passes the request provided
+// as a function parameter to the server.
+func newMockListener(endpoint  net.Conn) mockListener {
+
+	channel := make(chan net.Conn, 1)
+	channel <- endpoint
```
Comment (empijei) [inline http_server/request_testing/request_listener.go]: naming a variable after its type is not very telling, if the variable is short-lived you can shorten the name to be a single letter.

File: http_server/request_testing/request_listener.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	return listener
+}
+
+// Passes an endpoint to the server. Weirdly enough, when I make one request,
+// Accept() and hence Close() are being called twice, but the handler only once.
+func (l *mockListener) Accept() (net.Conn, error) {
+	log.Print("Accept() called")
+	ch, ok := <- l.connChannel
+	if !ok {
+		return nil, errors.New("EOF")
```
Comment (empijei) [inline http_server/request_testing/request_listener.go]: You can use more telling error strings, or you can use io.EOF as an error.

File: http_server/request_testing/request_listener.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+func (l *mockListener) Close() error {
+	log.Print("Close() called")
+	// My intuition here would have been to close both the endpoint and the
+	// channel. However, if I do that, I trigger a panic of "close of closed
+	// channel" when I attempt to close the channel. If I only close the // //
+	// channel, the program never terminates (I
+	// assume the server continues listening)
+	l.serverEndpoint.Close()
+	// close(l.connChannel)
```
Comment (empijei) [inline http_server/request_testing/request_listener.go]: you can avoid double closing the channel by using a [`sync.Once`](https://golang.org/pkg/sync/#Once) to protect the entire body of this function.

If you don't close the channel the Accept call will never return and you'll leak a goroutine.

File: http_server/request_testing/request_test.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"testing"
+)
+
+type handlerRecorder struct {
+	req         *http.Request
+	finishedReq chan bool
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: channels that are used to signal termination are normally named "doneSomething", in this case doneReq.

File: http_server/request_testing/request_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"testing"
+)
+
+type handlerRecorder struct {
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: This would normally be called recorderHandler (the more generic name to the right, as what you care about should be the first thing that you read).

File: http_server/request_testing/request_test.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	// log.Print(hr.req.URL.Query()["vegetable"])
+	hr.finishedReq <- true
+}
+
+func makeRequest(req string) *http.Request {
+	e1, e2 := net.Pipe()
+	listener := newMockListener(e2)
+	done := make(chan bool)
+	defer listener.Close()
+	handler := &handlerRecorder{req: nil, finishedReq: done}
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: you can use the `make` call inside the struct literal, you don't need to have a separate variable for it. Try to restrict the scope of your variable to the semantic block they live in (e.g. in this case you never use "done" again, so it is better to hide it, or a reader of the code might be wondering why you're saving this).

File: http_server/request_testing/request_test.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	return handler.req
+}
+
+func TestSetup(t *testing.T) {
+	req := makeRequest("GET / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+
+	if req == nil {
+		t.Error("Expected GET Request, received nil")
+	}
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: This cannot happen. If a request is not received you never send over the channel, so the call to `makeRequest` never returns. In order for this to manifest you'd need to add a timeout to your makeRequest call.

File: http_server/request_testing/request_test.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	handler := &handlerRecorder{req: nil, finishedReq: done}
+	server := &http.Server{Handler: handler}
+	go server.Serve(&listener)
+	defer server.Shutdown(context.Background())
+	e1.Write([]byte(req))
+	<-handler.finishedReq
+	return handler.req
+}
+
+func TestSetup(t *testing.T) {
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: You are just asserting on the method in this test, so maybe call it TestMethod?

File: http_server/request_testing/request_test.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	if req.Method != http.MethodGet {
+		t.Errorf("Expected %s, received %s", http.MethodGet, req.Method)
+	}
+}
+
+// Ensures Query() only returns a map of size one and verifies whether sending a
+// request with two values for the same key eturns a []string of length 2
+// containing the correct values 
+func TestQMultipleValsSameKey(t *testing.T) {
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: Test function names will show up if the test fails, so feel free to be explicit, TestMultipleQueryParametersWithSameKey is a perfectly valid name. Same for other tests.

File: http_server/request_testing/request_test.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	valueOne := "potatO"
+	valueTwo := "Tomato"
+	reqString := "GET /?vegetable=" + valueOne + "&vegetable=" + valueTwo + " HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n"
+	log.Print(reqString)
+	req := makeRequest(reqString)
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: Optional: if you are declaring a lot of variables you can use a `var` block instead of `:='

e.g.
```go
var (
  a = 1
  b = 2
)
```

File: http_server/request_testing/request_test.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+	queryParams := req.URL.Query()
+	if len(queryParams) != 1 {
+		t.Errorf("Expected one query parameter to have been received, got %d", len(queryParams))
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: It is common to write assertion messages like

"len(Query): got %d want %d"

Reads quicker and it is more standard.

Same below.

File: http_server/request_testing/request_test.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+			t.Errorf("Expected 2 query parameters, got %d", len(queryParams))
+		}
+
+	if len(queryParams["vegetable"]) != 1 || len(queryParams["Vegetable"]) != 1 {
+		t.Errorf("Expected one value for query parameter vegetables and one for Vegetables, got %d and %d", len(queryParams["vegetable"]), len(queryParams["Vegetable"]))
+	}
+}
+
+// Ensure keys and values that contain non-ASCII characters are parsed correctly
+func TestQRomanianCharacters(t *testing.T) {
```
Comment (empijei) [inline http_server/request_testing/request_test.go]: I would call this TestQueryUnicode.

File: http_server/request_testing/request_test.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  req = makeRequest(reqString)
+  
+  if  listLen := len(req.URL.Query()[key]); listLen != 1 {
+	  t.Errorf("Expected 1 value for key %s but got %d", key, listLen)
+  }
+
+
+}
+
+
```
Comment (empijei) [inline http_server/request_testing/request_test.go:175]: Maybe add one test with invalid unicode? (You can encode it with \xXX encoding)

**Review by empijei — CHANGES_REQUESTED** [review]

File: http_server/request_testing/request_listener.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+	"errors"
+	"log"
+)
+
+type mockListener struct {
+	connChannel chan net.Conn
+	serverEndpoint net.Conn
+}
+
+// For each request, we create an HTTP listener that passes the request provided
```
Comment (kele) [inline http_server/request_testing/request_listener.go]: nit: Document what the function actually does, not how one specific usage looks like.

File: http_server/request_testing/request_test.go [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+	if len(queryParams) != 1 {
+		t.Errorf("Expected one query parameter to have been received, got %d", len(queryParams))
+	}
+
+	vegetableParamValues := queryParams["vegetable"]
+	if len(vegetableParamValues) != 2 {
+		t.Errorf("Expected 2 query parameter values got %d", len(vegetableParamValues))
+	}
+	if vegetableParamValues[0] != valueOne || vegetableParamValues[1] != valueTwo {
+		t.Errorf("Expected "+valueOne+" and "+valueTwo+", got %s and %s", vegetableParamValues[0], vegetableParamValues[1])
```
Comment (mattiasgrenfeldt) [inline http_server/request_testing/request_test.go]: You can use format specifiers for the expected values instead of concatenating them. `"Expected %s and %s, got %s and %s"`

File: http_server/request_testing/request_test.go [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// Ensure different casing in  keys results in different query parameters
+func TestQKeysDifferentCasing(t *testing.T) {
+		req := makeRequest("GET /?vegetable=potato&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+		queryParams := req.URL.Query()
+		if len(queryParams) != 2 {
+			t.Errorf("Expected 2 query parameters, got %d", len(queryParams))
+		}
```
Comment (mattiasgrenfeldt) [inline http_server/request_testing/request_test.go]: Unindent one step

File: http_server/request_testing/request_test.go (hunk 7 above)
Comment (maramihali) [reply http_server/request_testing/request_test.go]: is it a good idea to use timeout?

File: http_server/request_testing/request_listener.go [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+	listener := mockListener{
+		connChannel: channel,
+		serverEndpoint: endpoint,
+	}
+	
+	return listener
+}
+
+// Passes an endpoint to the server. Weirdly enough, when I make one request,
+// Accept() and hence Close() are being called twice, but the handler only once.
```
Comment (maramihali) [inline http_server/request_testing/request_listener.go]: Do you have any idea why Accept() and Close() are called twice? This happens for every test and seems wasteful to me/

File: http_server/request_testing/request_listener.go [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+func (l *mockListener) Close() error {
+	log.Print("Close() called")
+	// My intuition here would have been to close both the endpoint and the
+	// channel. However, if I do that, I trigger a panic of "close of closed
+	// channel" when I attempt to close the channel. If I only close the // //
+	// channel, the program never terminates (I
+	// assume the server continues listening)
+	l.serverEndpoint.Close()
+	// close(l.connChannel)
+	return nil
```
Comment (maramihali) [inline http_server/request_testing/request_listener.go:46]: Should this return something better?

**Review by mattiasgrenfeldt — CHANGES_REQUESTED** [review]

File: http_server/request_testing/request_listener.go (hunk 17 above)
Comment (mattiasgrenfeldt) [reply http_server/request_testing/request_listener.go]: So the `http.Server` will call `listener.Accept` continously. It's up to Accept to block until there is a new connection or there is an error. When a new connection appears the server will send it of in a new go routine to handle the connection and the server will then continue to wait for more new connections by calling Accept again.

Close is called twice since you call `defer listener.Close()` and `defer server.Shutdown()`. The call to `server.Shutdown()` calls listener.Close() internally.

File: http_server/request_testing/request_listener.go [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+	ch, ok := <-l.connChannel
+	if !ok {
+		return nil, io.EOF
+	}
+	return ch, nil
+}
+
+func (l *mockListener) Close() error {
+	log.Print("Close() called")
+	o := &l.closeOnce
```
Comment (kele) [inline http_server/request_testing/request_listener.go:41]: Why do you need a pointer?

File: http_server/request_testing/request_listener.go (hunk 18 above)
Comment (kele) [reply http_server/request_testing/request_listener.go:46]: The error from `l.serverEndpoint` possibly.

**Review by kele — COMMENTED** [review]
Could you split the infrastructure from the test cases? Put the test cases in a separate PR.

Create a new branch with `git checkout -b` and just rewrite the history of this one.

File: http_server/request_testing/request_listener.go (hunk 19 above)
Comment (maramihali) [reply http_server/request_testing/request_listener.go:41]: As far as I understand, the Do() function expects a pointer to be passed
